File: src/PlatynUI/technology/uiautomation/core/mousedevice.py
# SPDX-FileCopyrightText: 2024 Daniel Biehl <daniel.biehl@imbus.de>
#
# SPDX-License-Identifier: Apache-2.0

from PlatynUI.core.types import Point, Size
from PlatynUI.ui.core.devices.basemousedevice import BaseMouseDevice
from PlatynUI.ui.core.devices.mousebutton import MouseButton

from .loader import DotNetInterface


class UiaMouseDevice(BaseMouseDevice):
    @property
    def double_click_time(self) -> float:
        return DotNetInterface.mouse_device().GetDoubleClickTime()

    @property
    def double_click_size(self) -> Size:
        r = DotNetInterface.mouse_device().GetDoubleClickSize()
        return Size(r.Width, r.Height)

    def get_position(self) -> Point:
        r = DotNetInterface.mouse_device().GetPosition()
        return Point(r.X, r.Y)

    def move_to(self, pos: Point) -> None:
        DotNetInterface.mouse_device().Move(pos.x, pos.y)

    def press(self, button: MouseButton) -> None:
        DotNetInterface.mouse_device().Press(int(button))

    def release(self, button: MouseButton) -> None:
        DotNetInterface.mouse_device().Release(int(button))

File: src/PlatynUI.Runtime/MouseDevice.cs
// SPDX-FileCopyrightText: 2024 Daniel Biehl <daniel.biehl@imbus.de>
//
// SPDX-License-Identifier: Apache-2.0

using System.ComponentModel.Composition;
using PlatynUI.Runtime.Core;

namespace PlatynUI.Runtime;

public class MouseDevice
{
    [Import]
    protected IMouseDevice? mouseDevice;

    private MouseDevice()
    {
        PlatynUiExtensions.ComposeParts(this);
    }

    private static MouseDevice? _instance;
    private static MouseDevice Instance => _instance ??= new MouseDevice();

    public static int GetDoubleClickTime()
    {
        return Instance.mouseDevice?.GetDoubleClickTime() ?? 0;
    }

    public static Size GetDoubleClickSize()
    {
        return Instance.mouseDevice?.GetDoubleClickSize() ?? new Size();
    }

    public static Point GetPosition()
    {
        return Instance.mouseDevice?.GetPosition() ?? new Point();
    }

    public static void Move(double x, double y)
    {
        Instance.mouseDevice?.Move(x, y);
    }

    public static void Press(MouseButton button)
    {
        Instance.mouseDevice?.Press(button);
    }

    public static void Press(int button)
    {
        Instance.mouseDevice?.Press((MouseButton)button);
    }

    public static void Release(MouseButton button)
    {
        Instance.mouseDevice?.Release(button);
    }

    public static void Release(int button)
    {
        Instance.mouseDevice?.Release((MouseButton)button);
    }
}
